show all numeric columns in singlestats/graph2

ryantxu opened the pull request:
This is a minor refactor to show all column results for single stats:

![image](https://user-images.githubusercontent.com/705951/54700168-ace77580-4aef-11e9-9328-e0a6109b389a.png)

This skips showing 'time' columns by adding a utility function that fills in the column type if it is not set explicitly

Also shows colums in the graph:
![image](https://user-images.githubusercontent.com/705951/54709763-584df580-4b03-11e9-84a2-58291e21141e.png)

ryantxu committed:
show all columns in singlestats

ryantxu committed:
cast to column

torkelo commented on packages/grafana-ui/src/utils/processTableData.ts:
```diff
@@ -4,6 +4,7 @@ import Papa, { ParseError, ParseMeta } from 'papaparse';
 
 // Types
 import { TableData, Column, TimeSeries } from '../types';
+import { isString } from 'util';
```
what is this? can't see a package dependency on util.

ryantxu replied:
weird -- not sure, autocomplete from vscode.  I switched to the lodash version.

torkelo commented on public/app/plugins/panel/singlestat2/SingleStatPanel.tsx:
```diff
@@ -24,13 +24,31 @@ export const getSingleStatValues = (props: PanelProps<SingleStatBaseOptions>): D
-    const value = stat !== 'name' ? series.stats[stat] : series.label;
-    return processor(value);
+  const values: DisplayValue[] = [];
+  data.forEach(t => {
```
prefer for of loops for all loops (10x more performant as well for large arrays as it compiles to native for loop)

ryantxu replied:
👍 good to know.  In this case that loop is on the list of tables, not the rows.  I changed it to for loop and will push soon

ryantxu committed:
incorporate torkels feedback

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into show-all-columns

* grafana/master:
  flot pairs

ryantxu committed:
show all colums in graph

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into show-all-columns

* grafana/master: (56 commits)
  another change that didn't come with earlier commit
  change that didn't come with in last commit
  reversed dashboard-padding
  Update CloudWatch metrics/dimension list (#16102)
  brought back dashboard-padding and panel-padding variables, made dashboard-padding more specific
  fix(prometheus): Change aligment of range queries (#16110)
  Minor refactoring of testdata query order PR #16122
  cleaner version
  maintain query order
  Update PLUGIN_DEV.md
  Merge with master, and updated logo and name
  more fixes to snapshot
  more fixes to snapshot
  removed empty space in snapshot
  fixed snapshot for test
  removed dashboard variables, removed headings-font-family variable, created theme variables for links and z-index, removed unused class in _panel_editor and _dashboard
  Tooltip: show percent instead of value
  Right tooltip position
  Add "No data points" message
  Improve tooltip look
  ...

torkelo commented on packages/grafana-ui/src/utils/processTableData.ts:
```diff
@@ -147,6 +149,67 @@ function convertTimeSeriesToTableData(timeSeries: TimeSeries): TableData {
+export const guessColumnTypes = (table: TableData): TableData => {
+  let changed = false;
+  const columns = table.columns.map((column, index) => {
+    if (!column.type) {
```
inverting this if case would be nice, as it would reduce the nesting

torkelo commented on packages/grafana-ui/src/utils/processTableData.ts:
```diff
@@ -147,6 +149,67 @@ function convertTimeSeriesToTableData(timeSeries: TimeSeries): TableData {
+/**
+ * @returns a table Returns a copy of the table with the best guess for each column type
+ */
+export const guessColumnTypes = (table: TableData): TableData => {
```
would be nice with 1-2 unit tests for this function.

torkelo commented on public/app/plugins/panel/graph2/GraphPanel.tsx:
```diff
@@ -19,12 +22,29 @@ export class GraphPanel extends PureComponent<Props> {
-        nullValueMode: NullValueMode.Ignore,
-      });
+    const vmSeries: TimeSeriesVMs = [];
+    for (let t = 0; t < data.length; t++) {
```
think can be simplified with, for (const table of data)

torkelo commented on public/app/plugins/panel/graph2/GraphPanel.tsx:
```diff
@@ -19,12 +22,29 @@ export class GraphPanel extends PureComponent<Props> {
+    for (let t = 0; t < data.length; t++) {
+      const table = guessColumnTypes(data[t]);
+      const timeColumn = getFirstTimeColumn(table);
+      if (timeColumn >= 0) {
```
nitpick: invert if condition, ie if (timeColumn <= 0) continue ... to reduce nesting

torkelo commented on public/app/plugins/panel/singlestat2/SingleStatPanel.tsx:
```diff
@@ -24,13 +24,31 @@ export const getSingleStatValues = (props: PanelProps<SingleStatBaseOptions>): D
-    return processor(value);
-  });
+  const values: DisplayValue[] = [];
+  for (let t = 0; t < data.length; t++) {
```
nitpick: replace with for of

torkelo reviewed (commented):
Looks great, mostly some nitpick comments :)

ryantxu committed:
cleanup and guess all columns

ryantxu committed:
drop one level of nesting

ryantxu commented:
@torkelo I took your nitpiks and made things more complicated :)  I think its better

Changes since you looked:
1. Rather than have each panel need to guess types, the `DataPanel` does it first and every panel will have them already set
2. I changed the `toTableData` utility so it converts a single object rather than an array.  DataPanel now handles the array processing
3. Guessing the type checks if a string can be converted to a number or boolean. 
4. Added a bunch of tests
5. Used `for( const x of ...` except where we need access to the numeric index

ryantxu committed:
removed option to not check strings
